### ISSUE TITLE
Why does a Cypher query involving relations require relation tables to exist?

### ISSUE THREAD

**Issue by ZmeiGorynych** [kuzudb/kuzu#3488]
When running a query such as [this](https://github.com/ShoggothAI/motleycrew/blob/148d95e36c13aa2ddbbf2ad4245e774e95c70223/motleycrew/applications/research_agent/question_task_recipe.py#L65C17-L67C18), that asserts that there are no relations of a certain kind connected to a given node, kuzu throws an exception if there are no relational tables at all.

Why should that be the case? We are working around that now by creating an empty, unrelated dummy relationship table, but this looks like a bug - if there are no relation tables, statements that involve relationships simply shouldn't match, rather than throwing. @whimo

**Comment by whimo** [comment]
This is even more general. I run this for an initially empty database:
```
CREATE NODE TABLE label1 (id SERIAL, PRIMARY KEY(id));
CREATE NODE TABLE label2 (id SERIAL, PRIMARY KEY(id));
CREATE (n:label1) RETURN n;
CREATE (n:label2) RETURN n;

MATCH (n:label1)-[r]->() WHERE n.id = 0 RETURN r;
// Binder exception: No rel table exists in database.

CREATE REL TABLE rel1 (FROM label1 TO label2);
MATCH (n:label1)-[r]->() WHERE n.id = 0 RETURN r;
// OK

MATCH (n:label1)-[r]->(k:label1) WHERE n.id = 0 RETURN r;
// Binder exception: Query node k violates schema. Expected labels are label2.

MATCH (n:label2)-[r]->() WHERE n.id = 0 RETURN r;
// Binder exception: Query node n violates schema. Expected labels are label1.
```

So, to query relationships from/to specific labels, there has to exist a relation table from/to exactly the same labels.

What's your view of this? It seems to me that much more logical would be to return empty results in such cases, because in many applications the node/relation tables are created on the go, and it should not affect the MATCH queries IMO.

**Comment by semihsalihoglu-uw** [comment]
This is a fair point though I can see the benefit of both erroring and not erroring. On the one side, if you have a database that does not have any relationship records and you're searching for relationship patterns, you're writing something against the schema you defined. So you might want the system to error. I know of cases when people write wrong queries during development that are guaranteed to return empty results, but are never notified by the system about this mistake. 

On the other side, I see that the singleton pattern `NOT (a)-[]->(b)` that you're using is a very natural pattern, so I see why you just want to get an empty result.

I don't think there is an obvious position to take here but I'm open to changing the behavior for now and return empty results on these. And if some other users demand that they want the system to error, we can debating having a configurable "strict mode" where people can choose their behavior.

We can provide a fix to this by the next release but probably sooner on dev releases. We need to see internally which intern can take this on and when.

**Comment by whimo** [comment]
> On the one side, if you have a database that does not have any relationship records and you're searching for relationship patterns, you're writing something against the schema you defined. So you might want the system to error.

I would totally agree with this if we were discussing relational DBs. With graph databases, though, the concept of schema is not as relevant. Moreover, in openCypher, which Kuzu claims to implement, such queries would always be valid and just return empty results.

**Comment by semihsalihoglu-uw** [comment]
Well, we have a clear position about how we want people to look at graph databases and particularly Kuzu, which is not as database that supports schema-less records. In fact, we are very explicit that graphs in Kuzu have schema. Therefore our data model is quite close to the relational model. One way to look at Kuzu's model is that it is the relational model with explicit identification of some records as nodes and some records as edges. We are particularly optimizing the system for cases where people have clear schema on their records but still want graph-based modeling so they can use a graph query language, specifically openCypher (with some variations). This gives users access to many good features of graph querying, though maybe not all the features offered by systems like Neo4j that adopt a more flexible model. But a priori declaration of schema is critical for our querying performance and scalability, even if it limits some other features we can offer.

**Comment by whimo** [comment]
Thanks for clarifying. Looking forward to the change of behavior regarding relationship querying!

P.S. It would be very nice to have a page in docs which describes in what aspects Kuzu's implementation of Cypher is different from openCypher, as well as your view on data models.

**Comment by andyfengHKU** [comment]
This should be fixed in #3506 and should be available in the nightly build version.
